time/timer.go

```diff
@@ -0,0 +1,649 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package gxtime encapsulates some golang.time functions
+package gxtime
+
+import (
+	"container/list"
+	"fmt"
+	"log"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+import (
+	uatomic "go.uber.org/atomic"
+)
+
+var (
+	ErrTimeChannelFull   = fmt.Errorf("timer channel full")
+	ErrTimeChannelClosed = fmt.Errorf("timer channel closed")
+)
+
+// Init initializes a default timer wheel
+func Init() {
+	defaultTimerWheelOnce.Do(func() {
+		defaultTimerWheel = NewTimerWheel()
+	})
+}
+
+// Now returns the current time.
+func Now() time.Time {
+	return defaultTimerWheel.Now()
+}
+
+////////////////////////////////////////////////
+// timer node
+////////////////////////////////////////////////
+
+var (
+	defaultTimerWheelOnce sync.Once
+	defaultTimerWheel     *TimerWheel
+	nextID                TimerID
+	curGxTime             = time.Now().UnixNano() // current goext time in nanoseconds
+)
+
+const (
+	maxMS     = 1000
+	maxSecond = 60
+	maxMinute = 60
+	maxHour   = 24
+	maxDay    = 31
+	// ticker interval不能设置到这种精度，
+	// 实际运行时ticker的时间间隔会在1.001ms上下浮动,
+	// 当ticker interval小于1ms的时候，会导致TimerWheel.hand
+	// 和timeWheel.inc不增长，造成时间错乱：例如本来
+	// 1.5s运行的函数在持续2.1s之后才被执行
+	// minDiff = 1.001 * MS
+	minDiff       = 10e6
+	maxTimerLevel = 5
+)
+
+func msNum(expire int64) int64     { return expire / int64(time.Millisecond) }
+func secondNum(expire int64) int64 { return expire / int64(time.Minute) }
+func minuteNum(expire int64) int64 { return expire / int64(time.Minute) }
+func hourNum(expire int64) int64   { return expire / int64(time.Hour) }
+func dayNum(expire int64) int64    { return expire / (maxHour * int64(time.Hour)) }
+
+// TimerFunc defines the time func.
+// if the return error is not nil, the related timer will be closed.
+type TimerFunc func(ID TimerID, expire time.Time, arg interface{}) error
+
+// TimerID is the id of a timer node
+type TimerID = uint64
+
+type timerNode struct {
```

Comment:
pls add comment for this struct and its fields, so that to get its functionality

Reply:
done